src/rp2_common/hardware_adc/include/hardware/adc.h

```diff
@@ -128,15 +128,24 @@ static inline void adc_set_temp_sensor_enabled(bool enable) {
  *
  *  Performs an ADC conversion, waits for the result, and then returns it.
  *
- * \return Result of the conversion.
+ * \return Result of the conversion or negative value on error.
  */
-static inline uint16_t adc_read(void) {
+static inline int32_t adc_read(void) {
     hw_set_bits(&adc_hw->cs, ADC_CS_START_ONCE_BITS);
+    
+    while ((adc_hw->cs & (ADC_CS_READY_BITS|ADC_CS_ERR_BITS|ADC_CS_ERR_STICKY_BITS))==0) {
+         tight_loop_contents();
+    }
 
-    while (!(adc_hw->cs & ADC_CS_READY_BITS))
-        tight_loop_contents();
-
-    return (uint16_t) adc_hw->result;
+    if ((adc_hw->cs & ADC_CS_READY_BITS)>0) {
+        return adc_hw->result & 0xfff;
+    }
+    if ((adc_hw->cs & ADC_CS_ERR_BITS)>0) {
+        return -ADC_CS_ERR_BITS;
```

Comment:
It's a bit random to return -ADC_CS_ERR_BITS. It would be better to return a "standard" error code? Maybe PICO_ERROR_GENERIC or PICO_ERROR_NO_DATA?

Reply:
I see your point, but if we doing this we can't distinguished what caused the error.

Use 
 PICO_ERROR_NO_DATA  
and additional, define a new code:
PICO_ERROR_ADC_STICKY_BIT